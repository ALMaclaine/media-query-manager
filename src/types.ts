export type BreakpointDims = Map<number, string>;
