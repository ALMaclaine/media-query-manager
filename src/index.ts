export * from './media-query-manager';
export type {BreakpointDims} from './types';
