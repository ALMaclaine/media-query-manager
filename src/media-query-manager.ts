import { ensureNotPxEm } from 'empxrem';

import  {BreakpointDims} from './types';

export class MediaQueryManager extends EventTarget {
  private _watchers: Array<MediaQueryList> = [];
  private _breaks: Array<number> = [];
  private _active: number = 0;
  private readonly _finalBreaks: number[] = [];

  private _handler = () => {
    for (let i = 0; i < this._watchers.length; i++) {
      if (this._watchers[i].matches) {
        this._active = this._breaks[i];
        break;
      }
    }
    this._changed();
  }

  constructor(breaks: Array<number>, baseFontSize: number = 16) {
    super();

    this._finalBreaks = [...breaks, 99999];
    for (const pt of this._finalBreaks) {
      const dim = ensureNotPxEm(pt, baseFontSize);
      const query = `(max-width: ${dim})`;
      const watcher = window.matchMedia(query);
      watcher.addEventListener('change', this._handler);
      this._watchers.push(watcher);
      this._breaks.push(pt);
    }
    this._handler();
  }

  private _changed() {
    this.dispatchEvent(new Event('change'));
  }

  public _destroy = () => this._watchers.forEach(e => e.removeEventListener('change', this._handler));

  get breaks(): number[] {
    return this._breaks
  }

  get active(): number {
    return this._active;
  }
}
