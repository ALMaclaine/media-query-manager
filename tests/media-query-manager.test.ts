import * as mediaQueryManager from '../src';

describe('media-query-manager', () => {
    it('Has Correct API', () => {
        const keys = Object.keys(mediaQueryManager);
        expect(keys.length).toBe(1);
        expect(keys).toMatchSnapshot();
    });
});
