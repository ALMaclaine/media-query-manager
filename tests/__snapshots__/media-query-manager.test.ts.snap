// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media-query-manager Has Correct API 1`] = `
Array [
  "MediaQueryManager",
]
`;
